From 13cf74cc9030c3757e1939be8c9a2f512500f5f7 Mon Sep 17 00:00:00 2001
From: Mike Looijmans <mike.looijmans@topic.nl>
Date: Mon, 23 Sep 2019 12:13:41 +0200
Subject: [PATCH 09/25] Pinmux WIFI or SD based on bootmode selection

Pinmux the SD1 controller dynamically based on the effective bootmode.
If the bootmode is set to SD boot, pinumux the SD card, otherwise switch
the SD1 interface to EMIO so the WiFi chip can be connected there.
---
 .../psu_init_gpl.c                            | 56 +++++++++++--------
 1 file changed, 34 insertions(+), 22 deletions(-)

diff --git a/board/topic/zynqmp/zynqmp-topic-miamimp-xilinx-xdp/psu_init_gpl.c b/board/topic/zynqmp/zynqmp-topic-miamimp-xilinx-xdp/psu_init_gpl.c
index 7aa70fd3b6..67a70d6632 100644
--- a/board/topic/zynqmp/zynqmp-topic-miamimp-xilinx-xdp/psu_init_gpl.c
+++ b/board/topic/zynqmp/zynqmp-topic-miamimp-xilinx-xdp/psu_init_gpl.c
@@ -13060,7 +13060,7 @@ unsigned long psu_mio_init_data(void)
     * Configures MIO Pin 45 peripheral interface mapping
     * (OFFSET, MASK, VALUE)      (0XFF1800B4, 0x000000FEU ,0x00000010U)
     */
-	PSU_Mask_Write(IOU_SLCR_MIO_PIN_45_OFFSET, 0x000000FEU, 0x00000010U);
+	PSU_Mask_Write(IOU_SLCR_MIO_PIN_45_OFFSET, 0x000000FEU, 0x00000000U);
 /*##################################################################### */
 
     /*
@@ -22337,6 +22337,37 @@ int psu_protection(void)
 	return 0;
 }
 
+static void psu_init_sdio_pinmux(void)
+{
+	/* Check currently active boot mode, this is "5" for SD card boot */
+	if ((Xil_In32(0xFF5E0200) & 0x7) != 0x5) {
+		/* SD not detected, pinmux the WiFi chip card */
+		PSU_Mask_Write(IOU_SLCR_MIO_PIN_45_OFFSET, 0x000000FEU, 0x00000000U);
+		PSU_Mask_Write(IOU_SLCR_MIO_PIN_46_OFFSET, 0x000000FEU, 0x00000000U);
+		PSU_Mask_Write(IOU_SLCR_MIO_PIN_47_OFFSET, 0x000000FEU, 0x00000000U);
+		PSU_Mask_Write(IOU_SLCR_MIO_PIN_48_OFFSET, 0x000000FEU, 0x00000000U);
+		PSU_Mask_Write(IOU_SLCR_MIO_PIN_49_OFFSET, 0x000000FEU, 0x00000000U);
+		PSU_Mask_Write(IOU_SLCR_MIO_PIN_50_OFFSET, 0x000000FEU, 0x00000000U);
+		PSU_Mask_Write(IOU_SLCR_MIO_PIN_51_OFFSET, 0x000000FEU, 0x00000000U);
+		/* Set feedback clock to EMIO */
+		PSU_Mask_Write(IOU_SLCR_SDIO_CLK_CTRL_OFFSET,0x00040000U, 0x00040000U);
+		/* Voltage support, fixed slot */
+		PSU_Mask_Write(IOU_SLCR_SD_CONFIG_REG2_OFFSET, 0x33840000U, 0x12800000U);
+	} else {
+		/* Set pin 45 as SD card detect, pin 46..51 to SDIO1 */
+		PSU_Mask_Write(IOU_SLCR_MIO_PIN_45_OFFSET, 0x000000FEU, 0x00000010U);
+		PSU_Mask_Write(IOU_SLCR_MIO_PIN_46_OFFSET, 0x000000FEU, 0x00000010U);
+		PSU_Mask_Write(IOU_SLCR_MIO_PIN_47_OFFSET, 0x000000FEU, 0x00000010U);
+		PSU_Mask_Write(IOU_SLCR_MIO_PIN_48_OFFSET, 0x000000FEU, 0x00000010U);
+		PSU_Mask_Write(IOU_SLCR_MIO_PIN_49_OFFSET, 0x000000FEU, 0x00000010U);
+		PSU_Mask_Write(IOU_SLCR_MIO_PIN_50_OFFSET, 0x000000FEU, 0x00000010U);
+		PSU_Mask_Write(IOU_SLCR_MIO_PIN_51_OFFSET, 0x000000FEU, 0x00000010U);
+		/* Set feedback clock to MIO 51 */
+		PSU_Mask_Write(IOU_SLCR_SDIO_CLK_CTRL_OFFSET, 0x00070000U, 0x00000000U);
+		/* Voltage support 3v3 only, removable slot */
+		PSU_Mask_Write(IOU_SLCR_SD_CONFIG_REG2_OFFSET, 0x33840000U, 0x00800000U);
+	}
+}
 
 int
 psu_init(void)
@@ -22357,29 +22388,10 @@ psu_init(void)
 	status &=    psu_afi_config();
 	psu_ddr_qos_init_data();
 
-	if (status == 0)
-		return 1;
-	return 0;
-}
-
-int psu_init_ddr_self_refresh(void) {
-
-    int status = 1;
-
-	status &= psu_mio_init_data();
-	status &=  psu_peripherals_pre_init_data();
-	status &=   psu_pll_init_data();
-	status &=   psu_clock_init_data();
-	status &=  psu_ddr_init_data();
-	status &=  psu_peripherals_init_data();
-	status &=  init_serdes();
-	init_peripheral();
-
-	status &=  psu_peripherals_powerdwn_data();
-	status &=    psu_afi_config();
-	psu_ddr_qos_init_data();
+	psu_init_sdio_pinmux();
 
 	if (status == 0)
 		return 1;
 	return 0;
 }
+
-- 
2.17.1

